### PR TITLE
Ilms 211 new design for recording control bar

### PR DESCRIPTION
**Pull request by Channa53**
### PR Info
- [x] Please functional test this PR
- [ ] Already tested in KMS and stand alone

### CODE REVIEW

File: src/components/App/ExpressRecorder.tsx [hunk 1]
```diff
@@ -447,9 +447,9 @@ export class ExpressRecorder extends Component<ExpressRecorderProps, State> {
...
         const { constraints, shareScreenOn } = this.state;
         if (
-            ((selectedCamera && constraints.video) || // check if video was turn on/off
+            ((selectedCamera && constraints.video) || // check if video has turned on/off
                 (!selectedCamera && !constraints.video)) &&
-            ((selectedAudio && constraints.audio) || // check if audio was turn on/off
+            ((selectedAudio && constraints.audio) || // check if audio has turned on/off
                 (!selectedAudio && !constraints.audio)) &&
             (!selectedCamera || // check if device id has been changed
                 (constraints.video &&
```
Comment (tomm1990) [inline src/components/App/ExpressRecorder.tsx:455]: it's easy to understand why you got confused with this condition.
it is horrible typing and should be refactored asap.

File: src/components/App/ExpressRecorder.tsx [hunk 2]
```diff
@@ -733,24 +718,43 @@ export class ExpressRecorder extends Component<ExpressRecorderProps, State> {
...
-                    )}
                     {doRecording && (
                         <RecordingTimer
                             onButtonClick={this.handleStopClick}
                             maxRecordingTime={maxRecordingTime}
                             setStopButtonRef={this.setStopButtonRef}
                         />
                     )}
+                    <div className={styles["settings-wrap"]}>
+                        {!doPlayback && !doRecording && !doCountdown && (
```
Comment (tomm1990) [inline src/components/App/ExpressRecorder.tsx:729]: I get it. you use to cut and paste the logic from one location to another. but is it hard to follow if the condition was met.
therefore, I suggest wrapping a variable that holds the answer to show the `<Settings />` or not.

please create a variable that holds your logic (I believe you should call him "isPlaying") and wrap it with **useMemo** hook

```
const isPlaying = useMemo(() => {
  if(isPlayback){
    return true;
  }
  if(){
    return ...
  }
  return false;
}, [isPlayback, ...]);

```

and after that, you can use:
```
<div className={styles["settings-wrap"]}>
{ !isPlaying ? <Settings /> : null }
</div>
```
once again. do not use `&&` for conditional rendering.

File: src/components/App/ExpressRecorder.tsx [hunk 3]
```diff
@@ -733,24 +718,43 @@ export class ExpressRecorder extends Component<ExpressRecorderProps, State> {
...
+                                selectedAudioDevice={
+                                    stream && stream.getAudioTracks().length > 0
+                                        ? ({
+                                              kind: "audioinput",
+                                              label: stream.getAudioTracks()[0].label
+                                          } as MediaDeviceInfo)
+                                        : false
+                                }
+                                allowVideo={allowVideo!}
+                                allowAudio={allowAudio!}
```
Comment (tomm1990) [inline src/components/App/ExpressRecorder.tsx:748]:
```suggestion
                                allowVideo={allowVideo}
```
why twice?
do not use `!` mark if you are not 100% sure what you are doing.
this mark is risky and might break the application.

File: src/components/App/style.scss [hunk 4]
```diff
@@ -86,9 +73,6 @@
 .settings-wrap {
     position: relative;
     display: inline-block;
-    float: right;
-    margin-right: 56px;
-    top: 16px;
-    color: #FFFFFF;
+    color: #111111;
```
Comment (tomm1990) [inline src/components/App/style.scss:76]: is there a const value for that color that you can use from other SCSS imports?

File: src/components/Playback/Playback.tsx [hunk 5]
```diff
@@ -75,7 +75,7 @@ export class Playback extends Component<Props, State> {
             });
             if (screenMedia) {
                 this.kalturaPlayerScreen = KalturaPlayer.setup({
-                    targetId: "player-wrap_screen" + uniqueId,
+                    targetId: "player-wrap-screen_" + uniqueId,
```
Comment (tomm1990) [inline src/components/Playback/Playback.tsx:78]:
```suggestion
                    targetId: "player-wrap_screen__" + uniqueId,
```
by BEM, you should use the underscore now. 
and since the `uniqueId` is unique, you should treat it like a special case (or state).

File: src/components/Playback/Playback.tsx [hunk 6]
```diff
@@ -110,7 +113,7 @@ export class Playback extends Component<Props, State> {
                 />
                 {screenMedia && (
                     <div
-                        id={"player-wrap_screen" + uniqueId}
+                        id={"player-wrap-screen_" + uniqueId}
```
Comment (tomm1990) [inline src/components/Playback/Playback.tsx:116]:
```suggestion
                        id={"player-wrap_screen__" + uniqueId}
```

File: src/components/Playback/player.css [hunk 7]
```diff
@@ -1,10 +1,13 @@
 .player-wrap__main_controls .playkit-bottom-bar,
 .player-wrap__main_controls .playkit-pre-playback-play-overlay {
-    left: 50% !important;
+    width: 200% !important;
```
Comment (tomm1990) [inline src/components/Playback/player.css:3]: would be great if we can avoid all of the `!important` in order to avoid future conflicts.
also, it's a bad practice.

File: src/components/Toggle-button/ToggleButton.tsx [hunk 8]
```diff
@@ -12,9 +12,7 @@ type Props = {
     onKeyPress?: () => void;
 };
 
-type State = {
-    isToggleOn: boolean;
-};
+type State = {};
```
Comment (tomm1990) [inline src/components/Toggle-button/ToggleButton.tsx:15]:
```suggestion
```
remove if not in use.

File: src/components/Settings/assets/video.svg [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<svg width="20" height="12" viewBox="0 0 20 12" fill="none" xmlns="http://www.w3.org/2000/svg">
+<path fill-rule="evenodd" clip-rule="evenodd" d="M2 0C0.895431 0 0 0.895431 0 2V10C0 11.1046 0.89543 12 2 12H11C12.1046 12 13 11.1046 13 10V2C13 0.895431 12.1046 0 11 0H2ZM15.4855 2.7087C15.1843 2.88942 15 3.21493 15 3.56619V8.43381C15 8.78507 15.1843 9.11058 15.4855 9.2913L18.4855 11.0913C19.152 11.4912 20 11.0111 20 10.2338V1.76619C20 0.988898 19.152 0.508785 18.4855 0.9087L15.4855 2.7087Z" fill="white"/>
+</svg>
```
Comment (tomm1990) [inline src/components/Settings/assets/video.svg:3]: please create a `JSX.component` that renders the content of this.
```
export const VideoIcon = (props) => {
  return <svg>...</svg>;
};

```
with this approach, we will be able to control the size, color, border (and more) of the vectors from within.
do not use direct SVGs like that.

File: src/components/Settings/Settings.tsx [hunk 10]
```diff
@@ -1,8 +1,13 @@
...
 import { SettingsDevices } from "./Settings-devices";
-import SettingsIcon from "./settings.svg";
 import { AudioIndicator } from "../AudioIndicator/AudioIndicator";
 import { Translator } from "../Translator/Translator";
+import VideoIcon from "./assets/video.svg";
+import AudioIcon from "./assets/audio.svg";
+import ScreenIcon from "./assets/screen.svg";
+import NoVideoIcon from "./assets/noVideo.svg";
+import NoAudioIcon from "./assets/noAudio.svg";
+import NoScreenIcon from "./assets/noScreen.svg";
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:10]: when importing icons, which are pretty much similar. I tend to import like:
`import { icon1, icon2, icon3 } from '../path/to/icons/folder'`

this will keep the code more organized

File: src/components/Settings/Settings.tsx [hunk 11]
```diff
@@ -11,21 +16,17 @@ type Props = {
         selectedAudio: MediaDeviceInfo | false,
         screenOn: boolean
     ) => void;
-    selectedCameraDevice?: MediaStreamTrack;
-    selectedAudioDevice?: MediaStreamTrack;
+    selectedCameraDevice: MediaDeviceInfo | false;
+    selectedAudioDevice: MediaDeviceInfo | false;
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:20]:
```suggestion
    selectedCameraDevice?: MediaDeviceInfo;
    selectedAudioDevice?: MediaDeviceInfo;
```
we should think about the term `false` here.
the condition that you wrote sounds like:
> selectedCameraDevice is MediaDeviceInfo OR false

which doesn't really seem like a reasonable logic for this variable.
if there is no `MediaDeviceInfo` at all, then is it `undefined` (and not `false`) - that sounds better and logic

File: src/components/Settings/Settings.tsx [hunk 12]
```diff
@@ -11,21 +16,17 @@ type Props = {
...
+    onStartRecording: () => void;
 };
 type State = {
-    isOpen: boolean;
-    selectedCamera: MediaDeviceInfo | false;
-    selectedAudio: MediaDeviceInfo | false;
-    showCameraSettings: boolean;
-    showAudioSettings: boolean;
-    showScreenShareSettings: boolean;
+    showSettingsOf: false | ResourceTypes;
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:29]:
```suggestion
    settingsType?: ResourceTypes;
```

**Review by tomm1990 — CHANGES_REQUESTED** [review]

File: src/components/Settings/Settings.tsx [hunk 13]
```diff
@@ -44,32 +45,12 @@ export class Settings extends Component<Props, State> {
...
-            };
-        }
         this.state = {
-            isOpen: false,
-            selectedCamera: selectedCamera,
-            selectedAudio: selectedAudio,
-            showAudioSettings: false,
-            showCameraSettings: false,
-            showScreenShareSettings: false,
+            showSettingsOf: false,
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:53]:
```suggestion
```
if no settings should be visible on the first render, then remove it from the `constructor`.

File: src/components/Settings/Settings.tsx [hunk 14]
```diff
@@ -163,69 +109,11 @@ export class Settings extends Component<Props, State> {
...
-
     handleClose = () => {
         document.removeEventListener("click", this.handleExternalClick, true);
 
         this.setState({
-            isOpen: false,
-            showAudioSettings: false,
-            showCameraSettings: false,
-            showScreenShareSettings: false
+            showSettingsOf: false
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:116]:
```suggestion
            this.setState((prevState) => {
              ...prevState,
              showSettingsOf: undefined
            });
```
to be on the safe side here, let's use the previous state and use a callback internal function to be excecuted.

File: src/components/Settings/Settings.tsx [hunk 15]
```diff
@@ -235,49 +123,48 @@ export class Settings extends Component<Props, State> {
...
-            }
-        );
-    };
+        this.setState((prevState: State) => {
+            return {
+                cameraOn: newCameraOn,
+                audioOn: newAudioOn,
+                screenOn: resourceType === ResourceTypes.SCREEN_SHARE ? isOn : prevState.screenOn
+            };
+        });
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:141]: please try to avoid those long conditions. it is really hard to understand them.
I suggest doing something like:
```
let { cameraOn } = this.state;

if(resourceType === ResourceTypes.VIDEO){
  cameraOn = isOn; // you can change this blocked scope variable, it is local
}
if(!isON && resourceType === ResourceTypes.AUDIO){
  cameraOn = true;
}
this.setState((prevState) => {
  return {
    ...prevState,
    cameraOn
  };
});
```

File: src/components/App/ExpressRecorder.tsx (hunk 2 above)
Comment (Channa53) [reply src/components/App/ExpressRecorder.tsx:729]: adding state variable for showing settings require managing 1 more state.. while I can conclude it from the others.. so I'm not sure it's better.

useMemo does not exist in preact. they recently added something similar and other hooks. I agree it's nice to have but it's require time - I would add hooks all over. @nireliyahu what do you think?

File: src/components/Settings/Settings.tsx [hunk 16]
```diff
@@ -235,49 +123,48 @@ export class Settings extends Component<Props, State> {
...
 
-    saveSettings = () => {
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:259]: since the settings have been changed, I don't know why you have removed this function.
is does exactly what is supposed to do.
any time you want to post to `this.props.onSettingsChanged`, you can just fire this function. 
I suggest to bring it back

File: src/components/App/ExpressRecorder.tsx (hunk 3 above)
Comment (Channa53) [reply src/components/App/ExpressRecorder.tsx:748]: haha you managed to confuse me.. it's  not twice - it's different props.
I removed the !

File: src/components/App/style.scss (hunk 4 above)
Comment (Channa53) [reply src/components/App/style.scss:76]: I'll keep this note till I get the final design colors.

File: src/components/Settings/Settings.tsx [hunk 17]
```diff
@@ -235,49 +123,48 @@ export class Settings extends Component<Props, State> {
...
         if (this.props.onSettingsChanged) {
-            const camera = this.state.cameraOn ? this.state.selectedCamera : false;
-            const audio = this.state.audioOn ? this.state.selectedAudio : false;
+            const camera = newCameraOn
+                ? this.props.selectedCameraDevice || this.cameraDevicesInfo[0]
+                : false;
+            const audio = newAudioOn
+                ? this.props.selectedAudioDevice || this.audioDevicesInfo[0]
+                : false;
             this.props.onSettingsChanged(camera, audio, this.state.screenOn);
```
Comment (tomm1990) [inline src/components/Settings/Settings.tsx:150]: please avoid that kind of hard-to-read conditions and split the entire logic into small pieces:
```
let { onSettingsChanged, selectedCameraDevice, selectedAudioDevice, onSettingsChanged } = this.props;
if(!onSettingsChanged){
  return;
}
if(cameraOn){
  selectedCameraDevice = selectedCameraDevice || this.cameraDevicesInfo[0];
}
...

onSettingsChanged(selectedCameraDevice, selectedAudioDevice, this.state.screenOn);
```

**Review by tomm1990 — CHANGES_REQUESTED** [review]

File: src/components/Playback/Playback.tsx (hunk 5 above)
Comment (Channa53) [reply src/components/Playback/Playback.tsx:78]: done

File: src/components/Playback/Playback.tsx (hunk 6 above)
Comment (Channa53) [reply src/components/Playback/Playback.tsx:116]: done

File: src/components/Playback/player.css (hunk 7 above)
Comment (Channa53) [reply src/components/Playback/player.css:3]: I know.. no choice here - I need to style player elements

File: src/components/Toggle-button/ToggleButton.tsx (hunk 8 above)
Comment (Channa53) [reply src/components/Toggle-button/ToggleButton.tsx:15]: in preact you have to have state object

File: src/components/Settings/assets/video.svg (hunk 9 above)
Comment (Channa53) [reply src/components/Settings/assets/video.svg:3]: in this case, with small project - I don't need to customize colors and sizes.. are there any other advanteges to use component?

File: src/components/Settings/Settings.tsx (hunk 11 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx]: done

File: src/components/Settings/Settings.tsx (hunk 12 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx:24]: done

File: src/components/Settings/Settings-devices.tsx [hunk 18]
```diff
@@ -73,28 +62,12 @@ export class SettingsDevices extends Component<Props, State> {
...
-    handleBack = () => {
-        this.props.onBack();
-    };
-
-    handleBackIconKeyPressed = (e: KeyboardEvent) => {
-        if (e.key === "Enter" || e.key === " ") {
-            this.handleBack();
+    handleToggleClick = () => {
+        if (this.props.onToggleChange) {
+            this.props.onToggleChange(!this.props.isOn);
```
[the remark this answers is not in the document]
Comment (Channa53) [reply src/components/Settings/Settings-devices.tsx:67]: done

File: src/components/Settings/Settings-devices.tsx [hunk 19]
```diff
@@ -139,39 +111,39 @@ export class SettingsDevices extends Component<Props, State> {
...
+                    {isSelected && (
+                        <span
+                            className={
+                                "device-label__selected-icon " +
+                                styles["device-label__selected-icon"]
+                            }
+                        >
+                            <CheckIcon />
+                        </span>
+                    )}
```
[the remark this answers is not in the document]
Comment (Channa53) [reply src/components/Settings/Settings-devices.tsx:129]: done

File: src/components/Settings/Settings-devices.tsx [hunk 20]
```diff
@@ -139,39 +111,39 @@ export class SettingsDevices extends Component<Props, State> {
...
-                </a>
+            <div className={`device-settings-wrap ${styles["device-settings-wrap"]}`}>
                 <ToggleButton
                     id={resourceName}
                     text={translator.translate(resourceName)}
                     onClick={this.handleToggleClick}
                     isToggleOn={isOn}
                     disabled={disabled}
-                    onKeyPress={this.handleKeyPress}
+                    onKeyPress={this.handleToggleClick}
```
[the remark this answers is not in the document]
Comment (Channa53) [reply src/components/Settings/Settings-devices.tsx:144]: onClick does not fire on key press, so we response for both

File: src/components/Settings/Settings-devices.tsx [hunk 21]
```diff
@@ -139,39 +111,39 @@ export class SettingsDevices extends Component<Props, State> {
...
                 />
-                <hr className={styles["settings-line"]} />
+                {resourcesList.length > 0 && <hr className={styles["settings-line"]} />}
```
[the remark this answers is not in the document]
Comment (Channa53) [reply src/components/Settings/Settings-devices.tsx:146]: done

File: src/components/Settings/Settings.tsx (hunk 15 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx:136]: it's needed.. the logic here need to consider the state, and new values. what you wrote is not equivalent to the code

File: src/components/Settings/Settings.tsx (hunk 13 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx:48]: done

File: src/components/Settings/Settings.tsx (hunk 16 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx:259]: I did not remove the logic - just united 2 functions into handle setting change

File: src/components/Settings/Settings.tsx (hunk 17 above)
Comment (Channa53) [reply src/components/Settings/Settings.tsx]: I find ? opertaor more readable then nested conditions, but I'' add variable decleration
